client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -126,111 +175,70 @@ public JwtTokenProviderImpl(String token, Controller controllerClient, String sc
-        }
-        return isTokenResetSuccessful;
+        return currentInstant.plusSeconds(refreshThresholdInSeconds).getEpochSecond() >= expiration.getEpochSecond();
     }
 
     @VisibleForTesting
-    Long extractExpirationTime(String token) {
-        if (token == null || token.trim().equals("")) {
-            return null;
+    String refreshToken() {
+        long traceEnterId = LoggerHelpers.traceEnter(log, "refreshToken", this.scopeName, this.streamName);
+        CompletableFuture<Void> currentRefreshFuture = tokenRefreshFuture.get();
+        if (currentRefreshFuture == null) {
+            log.debug("Initiated token refresh for scope {} and stream {}", this.scopeName, this.streamName);
+            currentRefreshFuture = this.recreateToken();
+            this.tokenRefreshFuture.compareAndSet(null, currentRefreshFuture);
+        } else {
+            log.debug("Token is already under refresh for scope {} and stream {}", this.scopeName, this.streamName);
         }
-        String[] tokenParts = token.split("\\.");
-
-        // JWT is of the format abc.def.ghe. The middle part is the body.
-        String encodedBody = tokenParts[1];
-        String decodedJsonBody = new String(Base64.getDecoder().decode(encodedBody));
-
-        return parseExpirationTime(decodedJsonBody);
+        try {
+            currentRefreshFuture.join(); // Block until the token is refreshed
+        } finally {
+            this.tokenRefreshFuture.compareAndSet(currentRefreshFuture, null); // Token is already refreshed, so resetting the future to null.
+        }
+        LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
+        return delegationToken.get().getValue();
     }
 
-    @VisibleForTesting
-    Long parseExpirationTime(String json) {
-        Long result = null;
-        if (json != null && !json.trim().equals("")) {
-            // Sample inputs:
-            //    {"sub":"subject","aud":"segmentstore","iat":1569837384,"exp":1569837434}
-            //    {"sub": "subject","aud": "segmentstore","iat": 1569837384,"exp": 1569837434}
-            Pattern pattern = Pattern.compile("\"exp\":\\s?(\\d+)");
-            Matcher matcher = pattern.matcher(json);
-            if (matcher.find()) {
-               // Will look like this, if a match is found: "exp": 1569837434
-               String matchedString = matcher.group();
-
-               String[] expiryTimeFieldParts = matchedString.split(":");
-               if (expiryTimeFieldParts != null && expiryTimeFieldParts.length == 2) {
-                   result = Long.parseLong(expiryTimeFieldParts[1].trim());
-               }
-            }
-        }
-        return result;
+    private CompletableFuture<Void> recreateToken() {
+        return controllerClient.getOrRefreshDelegationTokenFor(scopeName, streamName)
+                .thenApply(s -> new DelegationToken(s, extractExpirationTime(s)))
```

Comment:
Please combine these 2 callbacks into one. There's no need to add overhead when you can just have one callback that does both of these.

Reply:
Sure, I've collapsed the two callbacks into one as suggested.